Value domain assert fails when using np.nan values. np.isnan fails on reliability_data strings.

normangilmore opened the pull request:
The commit e2e565e9 adds the following assert:

```python
+            value_domain = np.asarray(value_domain)
+            assert np.isin(reliability_data, np.append(value_domain, np.nan)).all(), \
+                "The reliability data contains out-of-domain values."
```

This assert only works when object/string data is passed in for reliability_data and value_domain.
If nan is present in reliability_data, np.isin will return False. 

Example:
```python
import numpy as np
np.isin([1.,2.,3.,np.nan], np.append([1,2,3],np.nan)).all()
False
```

When I take the doctest:
```python
    >>> print(round(alpha(reliability_data, level_of_measurement='ordinal'), 3))
    0.815
```
and I add the value_domain parameter:
```python
    >>> print(round(alpha(reliability_data, value_domain=[1,2,3,4,5], level_of_measurement='ordinal'), 3))
    0.815

```

it returns
`AssertionError: The reliability data contains out-of-domain values.`

This is incorrect, and is due to np.nan being used in an equality test.

The following is a separate error that occurs when reliability_data is passed in as a string, and value_domain=None.
When I take the doctest:
```python
    >>> print(round(alpha(reliability_data, level_of_measurement='ordinal',
    ...                   value_domain=["very low", "low", "mid", "high", "very high"]), 3))
    0.815
```
and I remove the value_domain parameter:
```python
    >>> print(round(alpha(reliability_data, level_of_measurement='ordinal'), 3))
    0.815
```
it returns
`TypeError: ufunc 'isnan' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''`

This is caused by calling np.isnan on reliability_data when it is string/object data.
```python
        if value_domain is None:
            value_domain = np.unique(reliability_data[~np.isnan(reliability_data)])
        else:
            value_domain = np.asarray(value_domain)
```

I think there are too many policy decisions and interface decisions about how to support both string and float values for me to propose a complete fix. In any event, the parameters should work for float arrays always. I have pushed a commit removing the recently added assert, as that would be a breaking change for many/most users.

bryant1410 commented:
Thanks for the detailed PR!

I'm thinking the solution could be to have an auxiliary function that checks for NaNs even when the dtype is an object, then always use that one when checking for a NaN value.

Do you think you could make this change?

normangilmore commented:
Well, we know that np.nans or nulls are a legal and important value to allow in the reliability_data, including if string data is provided. So we don't really need to check for np.nan when verifying a provided value_domain.
I will give some thought to supporting these two additional test cases I have proposed.

bryant1410 commented:
Yeah but if they are not in the value domain then we need to check if they are NaN.

normangilmore commented:
I was thinking that we use np.unique to create a list of values and remove nans, as though we hadn't been given value_domain.

Then compare that to the value_domain passed in.

So with floats, this existing line works:
```python
value_domain = np.unique(reliability_data[~np.isnan(reliability_data)])
```

But this line doesn't work on strings. So we could just run np.unique, and then figure out a clean way to remove np.nan if the result array is a float type, and remove 'nan' if it is a string type. But hopefully in some way that doesn't involve looking at dtypes.

Maybe we can use a try/except block to try one way to filter out the nan, if that fails, assume it is a string, and try the other way.

bryant1410 commented:
I see. That sounds interesting!

To make sure it works with floats and strings, what do you think about having something similar to what we have now, but instead of doing `np.nan` (or checking for equality), we use an auxiliary function that checks for the NaN values that works for both strings (with equality) and floats (with `np.nan`)?

bryant1410 commented:
Hey! I was wondering if you have an ETA for this fix? Otherwise, maybe I can work on it.

createmomo commented:
> Hi guys,
> 
> I do have time to work on this finally. I need to refresh my memory and I will review and study this weekend.
> 
> I will probably have some follow up questions about how the API should work.
> 
> Norman

Thank you normangilmore! Hope this will be released soon!

normangilmore commented:
Hi @bryant1410,

I just pushed a branch with my proposed fixes for handling both numeric and string reliability_data, in particular when the value_domain is not provided and has to be inferred from the reliability_data. As part of validating value_domain, I added two ValueError exceptions for situations where the code can't determine a value_domain.

Note that we want to construct the found_value_domain regardless of whether it was provided or not, so we can either use it, or check the found_value_domain against the user provided value_domain as a sanity check.

When constructing the found_value_domain, I simply filter out np.nan or 'nan' depending on the dtype.kind. np.nan is not part of the value domain and is always allowed. Also, as we determined before, np.isin uses equality to test, and np.nan != np.nan. So removing np.nan/'nan' simplifies the assertion that checks the found_value_domain against the user provided value_domain.

I decided it was much cleaner and easier to follow the purpose of the code by examining dtype.kind rather than to try and catch TypeError exceptions which would be much more mysterious looking.

The two doctests I added previously now pass. The second doctest had to be changed to level_of_measurement='nominal' since it is not possible to infer an ordered value_domain from strings.

I rebased a squashed copy of this branch onto the latest main. Personally, I dislike merge commits and this repo has a very nice linear history. 😄  

It seems I used to know a trick for replacing the source branch on github with a new one, but my googling is failing me, so I'm opening a new PR# with a clean single commit on current main for this issue. See #20.

normangilmore commented:
@createmomo Sorry it took so long, but please see #20 and let me know if that resolves your issue.
